projects/pgai/pgai/vectorizer/generate/function_parser.py

```diff
@@ -0,0 +1,204 @@
+                                                         FROM pg_type at
+                                                         WHERE at.oid = t.typelem)
+                                ELSE t.typname
+                            END) as base_type,
+                           t.typtype = 'a' as is_array
+                    FROM pg_type t
+                    WHERE t.oid = %s
+                """
+                cur.execute(type_query, (type_oid,))
+                type_info = cur.fetchone()
+
+                type_name = type_info[1]  # type: ignore
+                is_array = type_info[2]  # type: ignore
+
+                default = None
+                # A parameter is required if it has no default value
+                is_required = i < non_default_count
+
+                parameters.append(
+                    PostgresParameter(
+                        name=name,  # type: ignore
+                        type_name=type_name,  # type: ignore
+                        default_value=default,
+                        is_array=is_array,  # type: ignore
+                        is_required=is_required,
+                    )
+                )
+
+            functions.append(
+                PostgresFunction(
+                    name=f"ai.{row[0]}",
+                    schema=row[1],
+                    parameters=parameters,
+                    return_type=row[7],
+                )
+            )
+
+    return functions
+
+
+@dataclass
+class VectorizerParameter:
+    """Dataclass to specifically handle create_vectorizer function parameters"""
+
+    name: str
+    python_type: str
+    accepted_configs: list[str] | None = None
+    is_required: bool = False
+    default_value: str | None = None
+
+
+def read_create_vectorizer_metadata(
+    conn: psycopg.Connection,
+) -> list[VectorizerParameter]:
+    """Extract and enhance metadata for the create_vectorizer function parameters"""
+    # First get basic function metadata
+    functions = get_function_metadata(conn, ["create_vectorizer"])
+    if not functions:
+        raise ValueError("create_vectorizer function not found")
+
+    vectorizer_fn = functions[0]
+
+    # Map parameter names to their enhanced metadata
+    parameters: list[VectorizerParameter] = []
+
+    # Query to get all configuration functions for each argument type
+    config_functions_query = """
+    SELECT p.proname
+    FROM pg_proc p
+    JOIN pg_namespace n ON p.pronamespace = n.oid
+    WHERE n.nspname = 'ai'
+    AND p.proname LIKE %s;
+    """
+
+    for param in vectorizer_fn.parameters:
+        # Skip internal parameters that shouldn't be exposed
+        if param.name.startswith("_"):
+            continue
+
+        accepted_configs = None
+        python_type = param.python_type
+
+        # For each parameter, check if there is corresponding config functions
+        with conn.cursor() as cur:
+            cur.execute(config_functions_query, (f"{param.name}_%",))
```

Comment:
I'm sure you're aware, but this has the assumption that the name of the configuring function contains the parameter name. I suspect we might trip over this at some point in the future.

Reply:
yeh that was a happy coincidence with our current function naming, if that's no longer the case, see the discussion about figuring out the function names in the first place, either we have to maintain a manual mapping in the generator or we figure out a way to automatically find the association between parameter and possible function out somehow.

Reply:
ack.